r55/src/exec.rs

```diff
@@ -385,3 +419,34 @@ fn dram_slice(emu: &mut Emulator, ret_offset: u64, ret_size: u64) -> Result<&mut
         Ok(&mut [])
     }
 }
+
+fn r55_gas_used(inst_count: &BTreeMap<String, u64>) -> u64 {
+    let total_cost = inst_count
+        .iter()
+        .map(|(instr_name, count)|
+            // Gas cost = number of instructions * cycles per instruction
+            match instr_name.as_str() {
+                // Gas map to approximate cost of each instruction
+                // References:
+                // http://ithare.com/infographics-operation-costs-in-cpu-clock-cycles/
+                // https://www.evm.codes/?fork=cancun#54
+                // Division and remainder
+                s if s.starts_with("div") || s.starts_with("rem") => count * 25,
+                // Multiplications
+                s if s.starts_with("mul") => count * 5,
+                // Loads
+                "lb" | "lh" | "lw" | "ld" | "lbu" | "lhu" | "lwu" => count * 3, // Cost analagous to `MLOAD`
+                // Stores
+                "sb" | "sh" | "sw" | "sd" | "sc.w" | "sc.d" => count * 3, // Cost analagous to `MSTORE`
+                // Branching
+                "beq" | "bne" | "blt" | "bge" | "bltu" | "bgeu" | "jal" | "jalr" => count * 3,
+                _ => *count, // All other instructions including `add` and `sub`
+        })
+        .sum::<u64>();
+
+    // This is the minimum "gas used" to ABI decode 'empty' calldata into Rust type arguments. Real calldata will take more gas.
+    // Internalising this would focus gas metering more on the function logic
+    let base_cost = 9_175_538;
```

Comment:
Whats base cost?

Reply:
This is the minimum cost to ABI decode calldata to Rust types, from [here](https://github.com/0xRampey/r55/blob/395b7732e74208afb932eece53c4ad16008c6695/contract-derive/src/lib.rs#L83)
Perhaps `base cost` is not the best naming and it could be `abi_decode_cost`. I've subtracted it because it overshadows function execution cost and happens before it.

Reply:
But do we need to count it separately here with a constant? Doesn't the decoding code also run in RISCV?

Reply:
I counted it separately mainly because the decoding logic consumes significantly more cycles compared to the user function logic. 
So _every_ contract call incurs a minimum decoding cost of ~9M cycles + the actual cost of contract execution- which is in the thousands and negligible by comparison due to the difference in magnitude.

The decoding logic does run inside RISCV tho and if we want to factor it in, we probably have to come up with a reduction like `0.001 * r55_cost`, to be able to properly add it to EVM gas costs.

Reply:
Yeah! I think we're g2g with this PR then